### ISSUE TITLE
【Feature Request】Will the codebse support relative position embedding ? 

### ISSUE THREAD

**Issue by kaixinbear** [HazyResearch/flash-attention#181]
Thanks for your great work. I wanna to adopt flash attention in my codebase but I find that it's hard to apply it to the attention with ralative position embedding. Could you help me?

**Comment by tridao** [comment]
The CUDA implementation doesn't support relative position embedding. The experimental Triton [implementation](https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attn_triton.py) has attention bias (and hence could work with relative pos emb) but I found the backward gradient to be sometimes wrong due to race conditions (from the Triton compiler).
Not sure how helpful this is but you can play with the Triton implementation.

**Comment by zijwang** [comment]
@tridao do you know whether (and when, if not) the triton implementation can support relative positional embeddings with learnable parameters, e.g., [kerple](https://openreview.net/forum?id=hXzOqPlXDwm&referrer=%5Bthe%20profile%20of%20Peter%20Ramadge%5D(%2Fprofile%3Fid%3D~Peter_Ramadge1))?

**Comment by tridao** [comment]
Relative pos emb can be done with attention bias. The experimental Triton implementation linked above supports attention bias (the forward seems robust enough, the backward might have some race conditions).
I don't have the bandwidth to work on this but you can play with the Triton code, it's a self-contained Python file.

**Comment by zijwang** [comment]
@tridao thanks for the reply. However, per my understanding, if we go with attenton bias for relative pos emb, then it cannot be learnable. Coudl you recommend a path if we wanna have learnable relative pos emb in triton flash attn?

**Comment by tridao** [comment]
> @tridao thanks for the reply. However, per my understanding, if we go with attenton bias for relative pos emb, then it cannot be learnable. Coudl you recommend a path if we wanna have learnable relative pos emb in triton flash attn?

Right, you'd need to implement the attention bias gradient in the backward pass. You can start with the Triton implementation and implement it there.
